File: Google-Kick-Start/2017/Kick_Start_2017-B-3.py
# Christmas Tree
# https://codingcompetitions.withgoogle.com/kickstart/round/0000000000201d27/0000000000201c9a
# DP, O(NMK*NM)

MIN_INT = -pow(10, 9)

T = int(input())
for t in range(T):
    [N, M, K] = [int(x) for x in input().split()]
    grid = []
    for _ in range(N):
        grid.append(input())

    prefix = [[0] for _ in range(N)]
    # prefix[i][j]: # of greens between grid[i][0] ~ grid[i][j], inclusive
    dp = [[[MIN_INT] * (K+1) for _ in range(M)] for _ in range(N+1)]
    # dp[i][j][k]: the max # of greens of k-tree starting at grid[i][j]
    for i in range(N+1):
        for j in range(M):
            dp[i][j][0] = 0

    for i in range(N):
        for j in range(M):
            prefix[i].append(prefix[i][-1] + int(grid[i][j] == '#'))
    # for i in range(N):
    #     print(prefix[i])
    
    for k in range(1, K+1):
        for i in range(N-1, -1, -1):
            for j in range(M):
                if grid[i][j] == '#':
                    for level in range(N - i):
                        row = i + level
                        l = j - level
                        r = j + level
                        if l < 0 or r >= M or prefix[row][r+1] - prefix[row][l] != r - l + 1:
                            break
                        tree = (level + 1) * (level + 1)
                        for x in range(l, r+1):
                            dp[i][j][k] = max(dp[i][j][k], tree + dp[row + 1][x][k - 1])
                        # print("i", i, "j", j, "k", k, "level", level, dp[i][j][k])

    ans = 0
    for i in range(N):
        for j in range(M):
            ans = max(ans, dp[i][j][K])

    print(f"Case #{t + 1}: {ans}")

File: Google-Kick-Start/2017/Kick_Start_2017-B-3.cpp
// Christmas Tree
// https://codingcompetitions.withgoogle.com/kickstart/round/0000000000201d27/0000000000201c9a
// DP, O(NMK*NM)

#include <bits/stdc++.h>

typedef long long ll;

using namespace std;

ll MIN_INT = -pow(10, 9);

int main() {
    int T;
    cin >> T;
    for (int t = 0; t < T; t++) {
        int N, M, K;
        cin >> N >> M >> K;
        vector<string> grid(N);
        for (int i = 0; i < N; i++) {
            cin >> grid[i];
        }

        vector< vector<int> > prefix(N, vector<int>(1, 0));
        // prefix[i][j]: # of greens between grid[i][0] ~ grid[i][j], inclusive
        vector< vector< vector<ll> > > dp(N+1, vector< vector<ll> >(M, vector<ll>(K+1, MIN_INT)));
        // dp[i][j][k]: the max # of greens of k-tree starting at grid[i][j]
        for (int i = 0; i <= N; i++) {
            for (int j = 0; j < M; j++) {
                dp[i][j][0] = 0;
            }
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                prefix[i].push_back(prefix[i].back() + (grid[i][j] == '#'));
            }
        }

        for (int k = 1; k <= K; k++) {
            for (int i = N - 1; i >= 0; i--) {
                for (int j = 0; j < M; j++) {
                    if (grid[i][j] == '#') {
                        for (int level = 0; level < N - i; level++) {
                            int row = i + level;
                            int l = j - level, r = j + level;
                            if (l < 0 || r >= M || prefix[row][r+1] - prefix[row][l] != r - l + 1)
                                break;
                            ll tree = (level + 1) * (level + 1);
                            for (int x = l; x <= r; x++)
                                dp[i][j][k] = max(dp[i][j][k], tree + dp[row + 1][x][k - 1]);
                        }
                    }
                }
            }
        }

        ll ans = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                ans = max(ans, dp[i][j][K]);
            }
        }
        
        printf("Case #%d: %lld\n", t + 1, ans);
    }
}
